cnetaddr_tostring_canonical_ipv6 test fails during `make check` on fresh clone of repo on macOS 10.14

**Expected behavior**
`make check` completes successfully when run on a freshly cloned [Bitcoin Core repo](https://github.com/bitcoin/bitcoin).

**Actual behavior**
`make check` fails with the following errors --

```
test/net_tests.cpp:446: error: in "net_tests/cnetaddr_tostring_canonical_ipv6": check net_addr.ToString() == expected_canonical_representation_output has failed [2001:db8::1:1:1:1:1 != 2001:db8:0:1:1:1:1:1]
test/net_tests.cpp:446: error: in "net_tests/cnetaddr_tostring_canonical_ipv6": check net_addr.ToString() == expected_canonical_representation_output has failed [2001:db8::1:1:1:1:1 != 2001:db8:0:1:1:1:1:1]
test/net_tests.cpp:446: error: in "net_tests/cnetaddr_tostring_canonical_ipv6": check net_addr.ToString() == expected_canonical_representation_output has failed [2001:db8::1:1:1:1:1 != 2001:db8:0:1:1:1:1:1]
test/net_tests.cpp:446: error: in "net_tests/cnetaddr_tostring_canonical_ipv6": check net_addr.ToString() == expected_canonical_representation_output has failed [2001:db8:aaaa:bbbb:cccc:dddd::1 != 2001:db8:aaaa:bbbb:cccc:dddd:0:1]
test/net_tests.cpp:446: error: in "net_tests/cnetaddr_tostring_canonical_ipv6": check net_addr.ToString() == expected_canonical_representation_output has failed [2001:db8:aaaa:bbbb:cccc:dddd::1 != 2001:db8:aaaa:bbbb:cccc:dddd:0:1]
...
...

Leaving test module "Bitcoin Core Test Suite"; testing time: 75790us

*** 5 failures are detected in the test module "Bitcoin Core Test Suite"
make[3]: *** [test/net_tests.cpp.test] Error 1
make[2]: *** [check-am] Error 2
make[1]: *** [check-recursive] Error 1
make: *** [check-recursive] Error 1
```

**To reproduce**
1. Follow the build steps available here -- [build-osx.md](https://github.com/bitcoin/bitcoin/blob/master/doc/build-osx.md) -- on a macOS 10.14.  Install all required dependencies and the first set of optional dependencies.  Use `Wallet (only SQlite) and GUI Support` configuration.
2. Proceed until the "2. Compile" step under the "Building Bitcoin Core" section, at which point running `make check` will result in the error above.

**System information**

Not running Core yet; this is a build time issue when running `make check` on the bitcoin repo cloned from https://github.com/bitcoin/bitcoin/ master branch at commit c8571486364d6e9ca8c86bd1c81e230ca64f8904.

```
$ ./src/bitcoind -version
Bitcoin Core version v21.99.0-1ef34ee25ed3
```

System Version:	macOS 10.14.6 (18G8022)
Processor Name:	Intel Core i5
  Processor Speed:	1.6 GHz
  Number of Processors:	1
  Total Number of Cores:	2
  L2 Cache (per Core):	256 KB
  L3 Cache:	4 MB
  Memory:	8 GB
File System:	APFS
Physical Drive:
  Device Name:	APPLE SSD AP0128N
  Media Name:	AppleAPFSMedia
  Medium Type:	SSD
  Protocol:	PCI-Express

--- a/src/netaddress.cpp
+++ b/src/netaddress.cpp
@@ -556,20 +556,57 @@ static std::string IPv4ToString(Span<const uint8_t> a)
     return strprintf("%u.%u.%u.%u", a[0], a[1], a[2], a[3]);
 }
 
+// Return an IPv6 address text representation with zero compression as described in RFC 5952
+// ("A Recommendation for IPv6 Address Text Representation").
 static std::string IPv6ToString(Span<const uint8_t> a)
 {
     assert(a.size() == ADDR_IPV6_SIZE);
-    // clang-format off
-    return strprintf("%x:%x:%x:%x:%x:%x:%x:%x",
-                     ReadBE16(&a[0]),
-                     ReadBE16(&a[2]),
-                     ReadBE16(&a[4]),
-                     ReadBE16(&a[6]),
-                     ReadBE16(&a[8]),
-                     ReadBE16(&a[10]),
-                     ReadBE16(&a[12]),
-                     ReadBE16(&a[14]));
-    // clang-format on
+    const std::array groups{
+        ReadBE16(&a[0]),
+        ReadBE16(&a[2]),
+        ReadBE16(&a[4]),
+        ReadBE16(&a[6]),
+        ReadBE16(&a[8]),
+        ReadBE16(&a[10]),
+        ReadBE16(&a[12]),
+        ReadBE16(&a[14]),
+    };
+
+    // The zero compression implementation is inspired by Rust's std::net::Ipv6Addr, see
+    // https://github.com/rust-lang/rust/blob/cc4103089f40a163f6d143f06359cba7043da29b/library/std/src/net/ip.rs#L1635-L1683
+    struct ZeroSpan {
+        size_t start_index{0};
+        size_t len{0};
+    };
+
+    // Find longest sequence of consecutive all-zero fields. Use first zero sequence if two or more
+    // zero sequences of equal length are found.
+    ZeroSpan longest, current;
+    for (size_t i{0}; i < groups.size(); ++i) {
+        if (groups[i] != 0) {
+            current = {i + 1, 0};
+            continue;
+        }
+        current.len += 1;
+        if (current.len > longest.len) {
+            longest = current;
+        }
+    }
+
+    std::string r;
+    r.reserve(39);
+    for (size_t i{0}; i < groups.size(); ++i) {
+        // Replace the longest sequence of consecutive all-zero fields with two colons ("::").
+        if (longest.len >= 2 && i >= longest.start_index && i < longest.start_index + longest.len) {
+            if (i == longest.start_index) {
+                r += "::";
+            }
+            continue;
+        }
+        r += strprintf("%s%x", ((!r.empty() && r.back() != ':') ? ":" : ""), groups[i]);
+    }
+
+    return r;
 }
 
 std::string CNetAddr::ToStringIP() const

--- a/src/netaddress.cpp
+++ b/src/netaddress.cpp
@@ -615,15 +615,6 @@ std::string CNetAddr::ToStringIP() const
     case NET_IPV4:
         return IPv4ToString(m_addr);
     case NET_IPV6: {
-        CService serv(*this, 0);
-        struct sockaddr_storage sockaddr;
-        socklen_t socklen = sizeof(sockaddr);
-        if (serv.GetSockAddr((struct sockaddr*)&sockaddr, &socklen)) {
-            char name[1025] = "";
-            if (!getnameinfo((const struct sockaddr*)&sockaddr, socklen, name,
-                             sizeof(name), nullptr, 0, NI_NUMERICHOST))
-                return std::string(name);
-        }
         return IPv6ToString(m_addr);
     }
     case NET_ONION:

